File: CivitasRuby/lib/mazo_sorpresas.rb
# encoding:utf-8
require_relative "sorpresa"
require_relative "diario"

module Civitas
  class MazoSorpresas
    
    def initialize (debug=false)
      @debug = debug
      init()
      if(debug)
        Diario.instance.ocurre_evento("debug on")
      end 
    end
    
    
    def init
      @sorpresas = Array.new
      @cartas_especiales = Array.new
      @ultima_sorpresa = nil
      @barajada = false
      @usadas = 0      
    end
    
    
    def al_mazo(s)
      if(!@barajada)
        @sorpresas << s
      end
    end
    
    
    def siguiente
      if(!@barajada || @usadas==@sorpresas.size() && !@debug)
        @sorpresas.shuffle
        @usadas=0
        @barajada = true
      end
      
      @usadas = @usadas+1
      @ultima_sorpresa = @sorpresas[0]
      @sorpresas.delete_at(0)
      @sorpresas << @ultima_sorpresa
      return @ultima_sorpresa 
    end
    
    
    def inhabilitar_carta_especial(sorpresa)  
      for i in @sorpresas
        if(i==sorpresa)
          @sorpresas.delete(i)
          @cartas_especiales << sorpresa
          Diario.instance.ocurre_evento("Carta inhabilitada")
        else
          Diario.instance.ocurre_evento("Carta no inhabilitada")
        end
      end
    end
    
    
    def habilitar_carta_especial(sorpresa)
      for i in @cartas_especiales
        if(i==sorpresa)
          @cartas_especiales.delete(sorpresa)
          @sorpresas << sorpresa          
          Diario.instance.ocurre_evento("Carta habilitada")
        else
          Diario.instance.ocurre_evento("Carta no habilitada")
        end
      end
    end
    
    private :init
    
  end
end

File: CivitasJava/src/Civitas/MazoSorpresas.java
package Civitas;
import java.util.ArrayList;
import java.util.Collections;

public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private Sorpresa ultimaSorpresa;
    
    private ArrayList<Sorpresa> cartasEspeciales;
    //static final private Diario diario = Diario.getInstance();
    
    
    private void init(){
        sorpresas = new ArrayList<>();
        cartasEspeciales = new ArrayList<>();
        barajada = false;
        usadas = 0;
    }
    
    MazoSorpresas(boolean estado){
        debug = estado;
        init();
        if(debug)
            Diario.getInstance().ocurreEvento("debug on");
    }
    
    MazoSorpresas(){
        debug = false;
        init();
    }
    
    void alMazo(Sorpresa s){
        if(!barajada)
            sorpresas.add(s);
    }
    
    Sorpresa siguiente(){
        if((!barajada || usadas == sorpresas.size()) && !debug ){
            Collections.shuffle(sorpresas);
            usadas = 0;
            barajada = true;
        }
        
        usadas++;
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(sorpresas.size(), ultimaSorpresa);
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        if(sorpresas.contains(sorpresa)){
            sorpresas.remove(sorpresa);
            cartasEspeciales.add(sorpresa);
            Diario.getInstance().ocurreEvento("Carta inhabilitada");
        }
        else
            Diario.getInstance().ocurreEvento("Carta no inhabilitada");
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        if(cartasEspeciales.contains(sorpresa)){
            cartasEspeciales.remove(sorpresa);
            sorpresas.add(sorpresas.size(), sorpresa);
            Diario.getInstance().ocurreEvento("Carta habilitada");
        }
        else
            Diario.getInstance().ocurreEvento("Carta no habilitada");
    }

    
}
